src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,784 @@
+
+            // Size of a tile.
+            m_spec.tile_width = static_cast<int>(props.m_tile_width);
+            m_spec.tile_height = static_cast<int>(props.m_tile_height);
+
+            m_spec.set_format(OIIO::TypeDesc::FLOAT);
+        }
+        
+        void write_tiles()
+        {
+            // Retrieve canvas properties.
+            const CanvasProperties& props = m_canvas->properties();
+
+            // Compute the tiles' xstride offset in bytes.
+            size_t xstride = props.m_pixel_size;
+
+            // Loop over the columns of tiles.
+            for (size_t tile_y = 0; tile_y < props.m_tile_count_y; tile_y++)
+            {
+                // Loop over the rows of tiles.
+                for (size_t tile_x = 0; tile_x < props.m_tile_count_x; tile_x++)
+                {
+                    // Compute the offset of the tile in pixels from the origin (origin: x=0;y=0).
+                    const size_t tile_offset_x = tile_x * props.m_tile_width;
+                    assert(tile_offset_x <= props.m_canvas_width);
+                    const size_t tile_offset_y = tile_y * props.m_tile_height;
+                    assert(tile_offset_y <= props.m_canvas_height);
+
+                    // Compute the tile's ystride offset in bytes.
+                    const size_t ystride = xstride * min(static_cast<size_t>(m_spec.width + m_spec.x - tile_offset_x),
+                        static_cast<size_t>(m_spec.tile_width));
+
+                    // Retrieve the (tile_x, tile_y) tile.
+                    const Tile& tile = m_canvas->tile(tile_x, tile_y);
+
+                    // Write the tile into the file.
+                    if (!m_writer->write_tile(
+                        static_cast<int>(tile_offset_x),
+                        static_cast<int>(tile_offset_y),
+                        0,
+                        OIIO::TypeDesc::FLOAT,
+                        tile.get_storage(),
+                        xstride,
+                        ystride))
+                    {
+                        const string msg = m_writer->geterror();
+                        close_file();
+                        throw ExceptionIOError(msg.c_str());
+                    }
+                }
+            }
+        }
+
+    private:
+        const ICanvas*      m_canvas;
+        OIIO::ImageSpec     m_spec;
+        OIIO::ImageOutput*  m_writer;
+        const char*         m_filename;
+    };
+
+    float hash_to_float(uint32_t hash) {
+        // if all exponent bits are 0 (subnormals, +zero, -zero) set exponent to 1
+        // if all exponent bits are 1 (NaNs, +inf, -inf) set exponent to 254
+        uint32_t exponent = hash >> 23 & 255; // extract exponent (8 bits)
+        if (exponent == 0 || exponent == 255)
+            hash ^= 1 << 23; // toggle bit
+        float f;
+        memcpy(&f, &hash, 4);
+        return f;
+    }
+}
+
+namespace renderer
+{
+
+namespace
+{
+
+    typedef map<uint32_t, string> NameMap;
```

Comment:
Include `<map>`.
Use `uint32` from `foundation` and include `foundation/platform/types.h`.